pkg/gardenlet/controller/bastion/reconciler.go

```diff
@@ -130,54 +125,90 @@ func (r *reconciler) reconcileBastion(
+
+			lastObservedError = fmt.Errorf("extension state is not Succeeded but %v", lastOperationState)
+			if extBastion.Status.LastError != nil {
+				lastObservedError = v1beta1helper.NewErrorWithCodes(fmt.Errorf("error during reconciliation: %s", extBastion.Status.LastError.Description), extBastion.Status.LastError.Codes...)
+			}
+		}
 	}
 
-	// wait for the extension controller to reconcile possible changes
-	if err := extensions.WaitUntilExtensionObjectReady(
-		ctx,
-		r.seedClient,
-		log,
-		extBastion,
-		extensionsv1alpha1.BastionResource,
-		defaultInterval,
-		defaultSevereThreshold,
-		defaultTimeout,
-		nil,
-	); err != nil {
-		if patchErr := patchReadyCondition(ctx, r.gardenClient, bastion, gardencorev1alpha1.ConditionFalse, "FailedReconciling", err.Error()); patchErr != nil {
+	if lastObservedError != nil {
+		message := fmt.Sprintf("Error while waiting for %s to become ready", extensionKey(extensionsv1alpha1.BastionResource, extBastion.Namespace, extBastion.Name))
+		err := gardencorev1beta1helper.NewErrorWithCodes(fmt.Errorf("%s: %w", message, lastObservedError), gardencorev1beta1helper.DeprecatedDetermineErrorCodes(lastObservedError)...)
+
+		if patchErr := patchReadyCondition(ctx, r.GardenClient, bastion, gardencorev1alpha1.ConditionFalse, "FailedReconciling", err.Error()); patchErr != nil {
 			log.Error(patchErr, "Failed patching ready condition")
 		}
+	}
+
+	if mustReconcileExtensionBastion {
+		if _, err := controllerutils.GetAndCreateOrMergePatch(ctx, r.SeedClient, extBastion, func() error {
+			metav1.SetMetaDataAnnotation(&extBastion.ObjectMeta, v1beta1constants.GardenerOperation, v1beta1constants.GardenerOperationReconcile)
+			metav1.SetMetaDataAnnotation(&extBastion.ObjectMeta, v1beta1constants.GardenerTimestamp, r.Clock.Now().UTC().String())
+
+			extBastion.Spec = extensionBastionSpec
+			return nil
+		}); err != nil {
+			if patchErr := patchReadyCondition(ctx, r.GardenClient, bastion, gardencorev1alpha1.ConditionFalse, "FailedReconciling", err.Error()); patchErr != nil {
+				log.Error(patchErr, "Failed patching ready condition")
+			}
 
-		return fmt.Errorf("failed wait for bastion extension resource to be reconciled: %w", err)
+			return fmt.Errorf("failed to ensure bastion extension resource: %w", err)
+		}
+		// return early here, the Bastion status will be updated by the reconciliation caused by the extension Bastion status update.
+		return nil
 	}
 
-	// copy over the extension's status to the garden and set the condition
-	patch := client.MergeFrom(bastion.DeepCopy())
-	setReadyCondition(bastion, gardencorev1alpha1.ConditionTrue, "SuccessfullyReconciled", "The bastion has been reconciled successfully.")
-	bastion.Status.Ingress = extBastion.Status.Ingress.DeepCopy()
-	bastion.Status.ObservedGeneration = &bastion.Generation
-	if err := r.gardenClient.Status().Patch(ctx, bastion, patch); err != nil {
-		return fmt.Errorf("failed patching ready condition of Bastion: %w", err)
+	if reconciliationSuccessful {
```

Comment:
Don't we need to also check if the lastOperationState is `gardencorev1beta1.LastOperationStateSucceeded`?